### PR TITLE
Amqp Value should be encoded properly when value was not initialized

### PR DESCRIPTION
**Pull request by Havret**
In nms-amqp we are using Message Encode/Decode methods to create message deep copy before we actually pass the message to the application. It is necessary in order to support session recovery scenario. (User shouldn't be able to mutate underlying message as it may be redelivered in the future).

https://github.com/apache/activemq-nms-amqp/blob/1a2d9b5cd27c25731b02568d7cf00cf92011e9cf/src/NMS.AMQP/Provider/Amqp/Message/AmqpNmsMessageFacade.cs#L430-L443

I found that there is a problem if you try to Encode/Decode message with AmqpValue payload, as it's not encoded properly if message value wasn't initialized before.

https://github.com/Azure/amqpnetlite/blob/abf0504cc9c6306e5da429d0e9eaec34873d97ae/src/Framing/AmqpValue.cs#L44-L54

If you do not explicitly call Value getter or setter, underlying value field will never be set. As it is later used to in EncodeValue we have a bug.

https://github.com/Azure/amqpnetlite/blob/abf0504cc9c6306e5da429d0e9eaec34873d97ae/src/Framing/AmqpValue.cs#L92

I've attached unit test illustrating the problem alongside with suggested solution. The problem only occur, when message arrive from the broker, if you try to encode message created in application it works fine, as value is already initialized. 

I'm not sure if I put test in the right place, but to illustrate the problem I need a broker to send a message "over the wire".

### CODE REVIEW

File: src/Framing/AmqpValue.cs [hunk 1]
```diff
@@ -89,7 +89,7 @@ internal override void EncodeValue(ByteBuffer buffer)
             }
             else
             {
-                this.WriteValue(buffer, this.value);
+                this.WriteValue(buffer, this.Value);
```
Comment (cezarypiatek) [inline src/Framing/AmqpValue.cs]: I think you should fix all the usages of `this.value`.  Without it, the condition above also will not work correctly.

**Comment by xinchen10** [comment]
The issue was caused by lazy decoding of a received message. The fix should work. There is a better solution by simply coping the valueBuffer into the encoding buffer in this particular case. It avoids decoding the wrapped value completely. When the body is a large byte[] or a collection of many items, it will be much more efficient.

**Comment by xinchen10** [comment]
Try this.
```
        internal override void EncodeValue(ByteBuffer buffer)
        {
            var byteBuffer = this.value as ByteBuffer;
            if (byteBuffer != null)
            {
                Encoder.WriteBinaryBuffer(buffer, byteBuffer);
            }
            else if (this.valueBuffer != null && !this.valueDecoded)
            {
                AmqpBitConverter.WriteBytes(buffer, this.valueBuffer.Buffer, this.valueBuffer.Offset, this.valueBuffer.Length);
            }
            else
            {
                this.WriteValue(buffer, this.value);
            }
        }
```

**Comment by HavretGC** [comment]
Great! It seems to work!

**Comment by Havret** [comment]
Great! It seems to work!

File: src/Framing/AmqpValue.cs [hunk 2]
```diff
@@ -82,11 +82,14 @@ protected virtual void WriteValue(ByteBuffer buffer, object value)
 
         internal override void EncodeValue(ByteBuffer buffer)
         {
-            var byteBuffer = this.value as ByteBuffer;
-            if (byteBuffer != null)
+            if (this.value is ByteBuffer byteBuffer)
```
Comment (xinchen10) [inline src/Framing/AmqpValue.cs]: Some of the projects for the old frameworks require VS 2015 to build. This new syntax is not supported, and we have to keep the old style of casting unfortunately.

Comment (Havret) [reply src/Framing/AmqpValue.cs]: Got it!

**Pull request merged**